### PR TITLE
Simplify condition variable use

### PR DESCRIPTION
**Pull request by heifner**

### CODE REVIEW

File: plugins/producer_plugin/producer_plugin.cpp [hunk 1]
```diff
@@ -419,18 +419,19 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
    
       // The queue storing read-only transactions to be executed by read-only threads
+      // Thread-safe
       class ro_trx_queue_t {
       public:
          void push_back(ro_trx_t&& trx) {
             std::unique_lock<std::mutex> g( mtx );
             queue.push_back(std::move(trx));
-            notify_waiting(false);
+            cond.notify_one();
```
Comment (greg7mdp) [inline plugins/producer_plugin/producer_plugin.cpp:428]:
```suggestion
            if (num_waiting)
               cond.notify_one();
```

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -473,34 +464,27 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          //    - all threads would be idle
          //    - or the net_plugin received a block.
          //    - or we have reached the read_window_deadline
-         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, fc::time_point deadline) {
-            assert(num_tasks > 0 && num_waiting == 0);
+         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, const fc::time_point& deadline) {
+            std::lock_guard<std::mutex> g( mtx );
```
Comment (greg7mdp) [inline plugins/producer_plugin/producer_plugin.cpp:468]: `std::lock_guard` unnecessary as this is called from a single thread

File: plugins/producer_plugin/producer_plugin.cpp [hunk 3]
```diff
@@ -473,34 +464,27 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          //    - all threads would be idle
          //    - or the net_plugin received a block.
          //    - or we have reached the read_window_deadline
-         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, fc::time_point deadline) {
-            assert(num_tasks > 0 && num_waiting == 0);
+         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, const fc::time_point& deadline) {
```
Comment (greg7mdp) [inline plugins/producer_plugin/producer_plugin.cpp:467]: I think it is better to pass deadline by value https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

File: plugins/producer_plugin/producer_plugin.cpp [hunk 4]
```diff
@@ -419,18 +419,19 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
             queue.push_back(std::move(trx));
-            notify_waiting(false);
+            cond.notify_one();
          }
          
          void push_front(ro_trx_t&& trx) {
             std::unique_lock<std::mutex> g( mtx );
             queue.push_front(std::move(trx));
-            notify_waiting(false);
+            cond.notify_one();
```
Comment (greg7mdp) [inline plugins/producer_plugin/producer_plugin.cpp:434]: same as above

**Comment by heifner** [comment]
Comments are being included into #844

File: plugins/producer_plugin/producer_plugin.cpp (hunk 1 above)
Comment (linh2931) [reply plugins/producer_plugin/producer_plugin.cpp:428]: Maybe we don't want to notify while in write window. `if (num_waiting)` can cover that.

**Pull request closed without merging**